File: Data Structures & Algorithms/house-robber/submission-7.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)

        if n <= 2:
            return max(nums)

        dp = [0] * n
        dp[-1], dp[-2] = nums[-1], nums[-2]

        for i in range(n - 3, -1, -1):
            dp[i] = max(dp[i + 1], nums[i] + dp[i + 2])

        return dp[0]

File: Data Structures & Algorithms/three-integer-sum/submission-7.cpp
#include <ranges>

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;

        for (const auto& [i, n] : views::enumerate(nums)) {
            if (i && n == nums[i - 1]) continue;
            int l = i + 1, r = nums.size() - 1;

            while (l < r) {
                int cur_sum = n + nums[l] + nums[r];

                if (cur_sum < 0) {
                    l++;
                } else if (cur_sum > 0) {
                    r--;
                } else {
                    res.push_back({n, nums[l], nums[r]});
                    l++; r--;

                    while (l < r && nums[l] == nums[l - 1]) {
                        l++;
                    }
                    while (l < r && nums[r] == nums[r + 1]) {
                        r--;
                    }
                }
            }
        }

        return res;
    }
};